### DISCUSSION TITLE
What js object should i use to store bytes in firestore using Admin / Web SDK?

### DISCUSSION THREAD

**Discussion by Vloz** [firebase/firebase-admin-node#2472] [Q&A] [answered]
[Already asked there without success](https://stackoverflow.com/questions/78012108/proper-type-to-store-bytes-in-firestore-js)

Js **Blob** and SDK **Bytes** return a serializing error.

`Error: Value for argument "data" is not a valid Firestore document. Couldn't serialize object of type "Bytes" (found in field "binaryData"). Firestore doesn't support JavaScript objects with custom prototypes (i.e. objects that were created via the "new" operator).`

Serializing works when i send it as Uint8Array but then it's displayed as a base64 string in firestore emulator..? 
And when i try to retrieve it with Web SDK it return to me as byteString/binaryString object..?

What's the correct way to store bytes in firestore?

**Comment by MILLERMARRU** [comment, marked as answer]
The correct type depends on which SDK you're using. Here's the full picture:

## Admin SDK (Node.js) — use `Buffer`

```ts
import { getFirestore } from 'firebase-admin/firestore';

const db = getFirestore();

// Storing
const buffer = Buffer.from('hello world', 'utf-8');
// From a file: const buffer = fs.readFileSync('./image.png');

await db.collection('files').doc('doc-id').set({
  data: buffer,  // Buffer is accepted directly by the Admin SDK
});

// Retrieving — comes back as Buffer on Node.js
const snap = await db.collection('files').doc('doc-id').get();
const retrieved: Buffer = snap.data()!.data;
console.log(retrieved.toString('utf-8')); // 'hello world'
```

## Web SDK (browser) — use `Bytes`

```ts
import { getFirestore, doc, setDoc, getDoc, Bytes } from 'firebase/firestore';

const db = getFirestore();

// Storing — wrap your Uint8Array in Bytes
const uint8 = new TextEncoder().encode('hello world');

await setDoc(doc(db, 'files', 'doc-id'), {
  data: Bytes.fromUint8Array(uint8),
  // or from base64: Bytes.fromBase64String('aGVsbG8gd29ybGQ=')
});

// Retrieving
const snap = await getDoc(doc(db, 'files', 'doc-id'));
const bytes: Bytes = snap.data()!.data;
const decoded = new TextDecoder().decode(bytes.toUint8Array());
console.log(decoded); // 'hello world'
```

## Why `Blob` and plain objects fail

Firestore only serializes its own special types plus native JS primitives. `Blob` has a custom prototype and is not on that list — hence the *"Firestore doesn't support JavaScript objects with custom prototypes"* error. `Uint8Array` works because the SDK internally converts it to the Firestore `Bytes` type.

## About the emulator showing base64

That's expected. The Firestore emulator UI displays `Bytes` fields as base64 strings — it's a display choice, not the actual stored format. When you retrieve the document in code, you get `Buffer` (Admin SDK) or `Bytes` (Web SDK), not a string.

**TL;DR:** `Buffer` with Admin SDK, `Bytes.fromUint8Array()` with Web SDK.
